pkg/build/metrics/prometheus/metrics.go

```diff
@@ -0,0 +1,172 @@
+)
+
+var (
+	// decided not to have a separate counter for failed builds, which have reasons,
+	// vs. the other "finished" builds phases, where the reason is not set
+	descterminalBuildCount = prometheus.NewDesc(
+		buildSubsystem+separator+terminalBuildCount,
+		"Counts total teriminal builds by namespace, name, phase, and in case of failures, reason",
+		[]string{"namespace", "name", "phase", "reason"},
+		nil,
+	)
+	descactiveBuildCount = prometheus.NewDesc(
+		buildSubsystem+separator+activeBuildCount,
+		"Show current counts of active builds by namespace, name, phase",
+		[]string{"namespace", "name", "phase"},
+		nil,
+	)
+	descbuildDuration = prometheus.NewDesc(
+		buildSubsystem+separator+buildDuration,
+		"Histogram of durations in seconds of terminated builds, by termination type",
+		[]string{"namespace", "name", "phase"},
+		nil,
+	)
+	descstageDuration = prometheus.NewDesc(
+		buildSubsystem+separator+stageDuration,
+		"Histogram of durations in seconds of stages that have finished, by stage name",
+		[]string{"namespace", "name", "stage"},
+		nil,
+	)
+	descstepDuration = prometheus.NewDesc(
+		buildSubsystem+separator+stepDuration,
+		"Histogram of durations in seconds of steps that have finished, by stage name and step name",
+		[]string{"namespace", "name", "stage", "step"},
+		nil,
+	)
+	bc         = buildCollector{}
+	registered = false
+	new        = strings.ToLower(string(buildapi.BuildPhaseNew))
+	pending    = strings.ToLower(string(buildapi.BuildPhasePending))
+	running    = strings.ToLower(string(buildapi.BuildPhaseRunning))
+)
+
+type buildCollector struct {
+	lister internalversion.BuildLister
+}
+
+// kube-state-metrics leverages prometheus.Registerer but does not exist in the version of prometheus
+// vendored into origin
+
+func IntializeMetricsCollector(buildLister internalversion.BuildLister) {
+	bc.lister = buildLister
+	// unit tests unearthed duplicate registration with prometheus via multiple calls to new build controller
+	if !registered {
+		prometheus.MustRegister(&bc)
+		registered = true
+	}
+	glog.V(4).Info("build metrics registered with prometheus")
+}
+
+// Describe implements the prometheus.Collector interface.
+func (bc *buildCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- descterminalBuildCount
+	ch <- descactiveBuildCount
+	ch <- descbuildDuration
+	ch <- descstageDuration
+	ch <- descstepDuration
+}
+
+// Collect implements the prometheus.Collector interface.
+func (bc *buildCollector) Collect(ch chan<- prometheus.Metric) {
+	result, err := bc.lister.List(kselector.Everything())
+
+	if err != nil {
+		glog.V(4).Infof("Collect err %v", err)
+		return
+	}
+
+	for _, b := range result {
+		bc.collectBuild(ch, b)
+	}
+
+	glog.V(4).Infof("reconcileActiveBuildMetrics looked at %d builds", len(result))
+
+}
+
+func addGauge(ch chan<- prometheus.Metric, b *buildapi.Build, desc *prometheus.Desc, v float64, lv ...string) {
+	lv = append([]string{b.ObjectMeta.Namespace, b.ObjectMeta.Name}, lv...)
```

Comment:
if we're trying to reduce allocations, we can reuse the label array, just changing the namespace, name, and state values, right?